design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,161 @@
+It is projected that COSI will be the only supported object storage driver in the future and OBCs will be deprecated after some time.
+
+## COSI Driver Deployment
+
+The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. It is not deployed by Rook.
+
+The Ceph COSI driver is deployed as a deployment with a [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver is deployed in the same namespace as the Rook operator. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:
+
+## Integration plan with Rook
+
+The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There will be one COSI Driver support from Rook. The driver will be started automatically with default settings when first Object Store gets created. The driver will be deleted when Rook Operator is uninstalled. The driver will be deployed in the same namespace as Rook operator. The user can provide additional setting via new `CephCOSIDriver` CRD which is owned by Rook.
+
+### Pre-requisites
+
+- COSI CRDs should be installed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
+```
+
+- COSI controller should be deployed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
+```
+
+### End to End Workflow
+
+#### Changes in Common.yaml
+
+Following contents need to be append to `common.yaml` :
+
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml>
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml>
+
+#### ceph-object-cosi-controller
+
+The `ceph-object-cosi-controller` will start the Ceph COSI Driver pod with default settings on the Rook Operator Namespace. The controller will bring up the Ceph COSI driver when first object store is created and will stop the COSI driver when last object store is deleted only if it detect COSI Controller is running on default namespace. The controller will also watch for CephCosiDriver CRD, if defined the driver will be started with the settings provided in the CRD. If the Ceph COSI driver if up and running,it  will also create `CephObjectStoreUser` named  `cosi` for each object store which internally creates a secret rook-ceph-object-user-<objectstore-name>-cosi provides credentials for the object store. This can be specified in the BucketClass and BucketAccessClass. Also this controller ensures maximum one CephCosiDriver CRD exists in the cluster. For v1.12 the Ceph COSI Driver will be supported only for single site CephObjectStore aka object stores not configured with multisite settings like zone/zonegroup/realm.
```

Comment:
Reposting since didn't see a response...

> The driver will be stopped when last Object Store is deleted.

I would propose to follow the same pattern for uninstall as we do for the CSI driver. For CSI, an ownerReference is added to its resources that points to the Rook operator deployment. So the CSI driver is deleted whenever the Rook operator deployment is uninstalled. The same behavior seems best for the COSI driver, then Rook doesn't need any cleanup code for the COSI driver and its lifetime is the same as the operator.